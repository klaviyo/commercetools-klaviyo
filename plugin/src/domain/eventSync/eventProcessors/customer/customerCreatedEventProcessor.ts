import { AbstractEventProcessor } from '../abstractEventProcessor';
import logger from '../../../../utils/log';
import { Customer, CustomerCreatedMessage } from '@commercetools/platform-sdk';

export class CustomerCreatedEventProcessor extends AbstractEventProcessor {
    isEventValid(): boolean {
        const customerCreatedMessage = this.ctMessage as unknown as CustomerCreatedMessage;
        return (
            customerCreatedMessage.resource.typeId === 'customer' &&
            customerCreatedMessage.type === 'CustomerCreated' &&
            !!customerCreatedMessage.customer &&
            !!customerCreatedMessage.customer.email
        );
    }

    async generateKlaviyoEvents(): Promise<KlaviyoEvent[]> {
        const message = this.ctMessage as unknown as CustomerCreatedMessage;
        logger.info(`processing CT ${message.resource.typeId}${message.type} message`);

        let customer: Customer;
        if ('customer' in message) {
            customer = message.customer;
        } else {
            customer = (await this.context.ctCustomerService.getCustomerProfile(
                (message as CustomerCreatedMessage).resource.id,
            )) as Customer;
        }

        const klaviyoEvent: KlaviyoEvent = {
            body: this.context.customerMapper.mapCtCustomerToKlaviyoProfile(customer),
            type: 'profileCreated',
        };
        return [klaviyoEvent];
    }
}
