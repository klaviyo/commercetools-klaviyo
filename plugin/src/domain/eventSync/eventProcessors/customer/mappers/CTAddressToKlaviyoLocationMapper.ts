import { Address } from '@commercetools/platform-sdk';

export const mapCTAddressToKlaviyoLocation = (address?: Address): KlaviyoLocation | null => {
    return address
        ? {
              address1: getAddressLine1(address),
              address2: getAddressLine2(address),
              city: address.city,
              country: address.country,
              region: address.region,
              zip: address.postalCode,
          }
        : null;
};

const getAddressLine1 = (address?: Address): string | null => {
    return address
        ? [address.apartment, address.building, address.streetNumber, address.streetName]
              .filter((element) => element)
              .join(', ')
        : null;
};

const getAddressLine2 = (address?: Address): string | null => {
    return address
        ? [address.additionalStreetInfo, address.additionalAddressInfo].filter((element) => element).join(', ')
        : null;
};
