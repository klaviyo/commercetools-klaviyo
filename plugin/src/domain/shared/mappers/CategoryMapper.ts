import { Category } from '@commercetools/platform-sdk';

export interface CategoryMapper {
    mapCtCategoryToKlaviyoCategory(category: Category, klaviyoCategoryId?: string): CategoryRequest;
    mapKlaviyoCategoryIdToDeleteCategoryRequest(klaviyoCategoryId: string): CategoryDeletedRequest;
}
