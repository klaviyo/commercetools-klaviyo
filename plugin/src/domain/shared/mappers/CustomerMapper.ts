import { Address, Customer } from '@commercetools/platform-sdk';

export interface CustomerMapper {
    mapCtCustomerToKlaviyoProfile(customer: Customer, klaviyoProfileId?: string): ProfileRequest;
    mapCTAddressToKlaviyoLocation(address?: Address, useSpecialPrefix?: boolean): KlaviyoLocation | null;
}
