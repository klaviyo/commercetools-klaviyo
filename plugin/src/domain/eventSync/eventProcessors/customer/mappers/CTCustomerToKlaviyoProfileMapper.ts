import { Customer } from '@commercetools/platform-sdk';
import { getCTCustomerAddressForKlaviyo, getPhoneNumber } from '../utils/CustomerAddressUtils';
import { mapCTAddressToKlaviyoLocation } from './CTAddressToKlaviyoLocationMapper';

/**
 * Maps a commercetools customer to the Klaviyo profile request attributes
 * @param customer
 */
export const mapCTCustomerToKlaviyoProfile = (customer: Customer): Profile => {
    const address = getCTCustomerAddressForKlaviyo(customer);
    const {
        email,
        firstName: first_name,
        lastName: last_name,
        title,
        companyName: organization,
        id: external_id,
        custom,
    } = customer;

    return {
        email,
        external_id,
        first_name,
        last_name,
        title,
        phone_number: getPhoneNumber(address),
        organization,
        location: mapCTAddressToKlaviyoLocation(address),
        properties: custom?.fields,
    };
};
