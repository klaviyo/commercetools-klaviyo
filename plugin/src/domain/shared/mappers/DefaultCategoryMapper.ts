import { Category } from '@commercetools/platform-sdk';
import { CategoryMapper } from './CategoryMapper';
import { getLocalizedStringAsText } from '../../../utils/locale-currency-utils';

export class DefaultCategoryMapper implements CategoryMapper {
    public mapCtCategoryToKlaviyoCategory(category: Category, klaviyoCategoryId?: string): CategoryRequest {
        return {
            data: {
                type: 'catalog-category',
                id: klaviyoCategoryId,
                attributes: {
                    external_id: !klaviyoCategoryId ? category.id : undefined,
                    name: this.getCategoryNameWithAncestors(category),
                    integration_type: !klaviyoCategoryId ? '$custom' : undefined,
                    catalog_type: !klaviyoCategoryId ? '$default' : undefined,
                },
            },
        };
    }

    private getCategoryNameWithAncestors(category: Category): string {
        const categoryNames = category.ancestors
            .map((ancestor) => getLocalizedStringAsText((ancestor.obj as Category).name))
            .concat(getLocalizedStringAsText(category.name));
        return categoryNames.join(' > ');
    }

    public mapKlaviyoCategoryIdToDeleteCategoryRequest(klaviyoCategoryId: string): CategoryDeletedRequest {
        return {
            data: {
                id: klaviyoCategoryId,
            },
        };
    }
}
