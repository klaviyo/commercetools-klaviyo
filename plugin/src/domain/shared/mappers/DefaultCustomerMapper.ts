import { Address, Customer } from '@commercetools/platform-sdk';
import { CustomerMapper } from './CustomerMapper';
import { mapAllowedProperties } from '../../../utils/property-mapper';

const E_164_REGEX = /^\+[1-9]\d{10,14}$/;

export class DefaultCustomerMapper implements CustomerMapper {
    public mapCtCustomerToKlaviyoProfile(customer: Customer, klaviyoProfileId?: string): ProfileRequest {
        const address = this.getCTCustomerAddressForKlaviyo(customer);
        const {
            email,
            firstName: first_name,
            lastName: last_name,
            title,
            companyName: organization,
            id: external_id,
            custom,
        } = customer;
        const props = mapAllowedProperties('customer.customFields', { ...(custom?.fields || {}) });

        return {
            data: {
                type: 'profile',
                id: klaviyoProfileId,
                attributes: {
                    email,
                    external_id,
                    first_name,
                    last_name,
                    title,
                    phone_number: address?.mobile || address?.phone,
                    organization,
                    location: this.mapCTAddressToKlaviyoLocation(address),
                    properties: Object.keys(props).length ? { ...props } : undefined,
                },
            },
        };
    }

    public mapCTAddressToKlaviyoLocation(address?: Address, useSpecialPrefix = false): KlaviyoLocation | null {
        const specialPrefix = useSpecialPrefix ? '$' : '';
        return address
            ? {
                  [`${specialPrefix}address1`]: this.getAddressLine1(address),
                  [`${specialPrefix}address2`]: this.getAddressLine2(address),
                  [`${specialPrefix}city`]: address.city,
                  [`${specialPrefix}country`]: address.country,
                  [`${specialPrefix}region`]: address.region,
                  [`${specialPrefix}zip`]: address.postalCode,
              }
            : null;
    }

    public getAddressLine1(address?: Address): string | null{
        return address
            ? [address.apartment, address.building, address.streetNumber, address.streetName]
                  .filter((element) => element)
                  .join(', ')
            : null;
    }

    public getAddressLine2(address?: Address): string | null {
        return address
            ? [address.additionalStreetInfo, address.additionalAddressInfo].filter((element) => element).join(', ')
            : null;
    }

    // No longer in use, validation only handled by Klaviyo.
    // Phone number removed in worse case scenario.
    public getPhoneNumber(address?: Address): string | null {
        const phoneNumber = address?.mobile || address?.phone;
        if (phoneNumber) {
            const noSpacesAndDashPhoneNumber = phoneNumber.replace(/\s|-|\/|\)|\(|\*|[a-z]|[A-Z]+/g, '');
            return E_164_REGEX.test(noSpacesAndDashPhoneNumber) ? noSpacesAndDashPhoneNumber : null;
        }
        return null;
    }

    /**
     * A CT customer can have multiple addresses, klaviyo accepts only one address information in the location field.
     * This method selects one of the CT addresses to be used to populate the klaviyo location info for a klaviyo profile.
     * The address selection evaluates in order the following conditions and returns the first address found:
     * 1. Selects the customer defaultBillingAddress if set
     * 2. Selects the first available billing address if present
     * 3. Selects the first available address if present
     * @param customer
     */
    public getCTCustomerAddressForKlaviyo(customer: Customer): Address | undefined {
        if (!customer || !customer.addresses.length) {
            return;
        }
        if (customer.defaultBillingAddressId) {
            return customer.addresses?.find((address) => address.id === customer.defaultBillingAddressId);
        }
        if (customer?.billingAddressIds?.length) {
            return customer.addresses?.find((address) => address.id === customer?.billingAddressIds?.find(() => true));
        }
        return customer.addresses?.find(() => true);
    }
}
