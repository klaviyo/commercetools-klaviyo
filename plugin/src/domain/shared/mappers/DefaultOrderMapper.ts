import { getCustomerProfileFromOrder } from '../../../utils/get-customer-profile-from-order';
import { getTypedMoneyAsNumber } from '../../../utils/get-typed-money-as-number';
import { mapAllowedProperties } from '../../../utils/property-mapper';
import { Category, CategoryReference, LineItem, Order, Product } from '@commercetools/platform-sdk';
import { OrderMapper } from './OrderMapper';
import config from 'config';
import { CurrencyService } from '../services/CurrencyService';
import { CustomerMapper } from './CustomerMapper';
import { getLocalizedStringAsText } from '../../../utils/locale-currency-utils';

export class DefaultOrderMapper implements OrderMapper {
    constructor(private readonly currencyService: CurrencyService, private readonly customerMapper: CustomerMapper) {}
    public mapCtOrderToKlaviyoEvent(
        order: Order,
        orderProducts: Product[],
        metric: string,
        updateAdditionalProfileProperties: boolean,
        time?: string,
    ): EventRequest {
        return {
            data: {
                type: 'event',
                attributes: {
                    profile: getCustomerProfileFromOrder(
                        order,
                        this.customerMapper,
                        updateAdditionalProfileProperties && metric === config.get('order.metrics.placedOrder'),
                    ),
                    metric: {
                        name: metric,
                    },
                    value: this.currencyService.convert(
                        getTypedMoneyAsNumber(order?.totalPrice),
                        order.totalPrice.currencyCode,
                    ),
                    properties: {
                        ...mapAllowedProperties('order', { ...order }),
                        ...mapAllowedProperties('order.customFields', { ...(order.custom?.fields || {}) }),
                        ItemNames: this.mapOrderLineItemsToItemNames(order),
                        Categories: this.getCategoryNamesFromProduct(
                            orderProducts.map((product) => product.masterData.current.categories).flat(),
                        ),
                    } as any,
                    unique_id: order.id,
                    time: time ?? order.createdAt,
                },
            },
        };
    }

    public mapCtRefundedOrderToKlaviyoEvent(
        order: Order,
        orderProducts: Product[],
        metric: string,
        time?: string,
    ): EventRequest {
        const refundAmounts =
            order.paymentInfo?.payments
                .map((p) =>
                    p.obj?.transactions
                        .filter((t) => t.state === 'Success' && t.type === 'Refund')
                        .map((t) => getTypedMoneyAsNumber(t.amount)),
                )
                .flat()
                .filter((r) => r !== undefined) || [];
        const refundTotal = refundAmounts.length ? refundAmounts.reduce((a, b) => (a || 0) + (b || 0)) || 0 : 0;

        return {
            data: {
                type: 'event',
                attributes: {
                    profile: getCustomerProfileFromOrder(order, this.customerMapper),
                    metric: {
                        name: metric,
                    },
                    value: this.currencyService.convert(refundTotal, order.totalPrice.currencyCode),
                    properties: {
                        ...mapAllowedProperties('order', { ...order }),
                        ...mapAllowedProperties('order.customFields', { ...(order.custom?.fields || {}) }),
                        ItemNames: this.mapOrderLineItemsToItemNames(order),
                        Categories: this.getCategoryNamesFromProduct(
                            orderProducts.map((product) => product.masterData.current.categories).flat(),
                        ),
                    } as any,
                    unique_id: order.id,
                    time: time ?? order.createdAt,
                },
            },
        };
    }

    public mapOrderLineToProductOrderedEvent(lineItem: LineItem, order: Order, time?: string): EventRequest {
        return {
            data: {
                type: 'event',
                attributes: {
                    profile: getCustomerProfileFromOrder(order, this.customerMapper),
                    metric: {
                        name: config.get('order.metrics.orderedProduct'),
                    },
                    value: this.currencyService.convert(
                        getTypedMoneyAsNumber(lineItem.totalPrice),
                        order.totalPrice.currencyCode,
                    ),
                    properties: { ...lineItem },
                    unique_id: lineItem.id,
                    time: time ?? order.createdAt,
                },
            },
        };
    }

    private mapOrderLineItemsToItemNames(order: Order): string[] {
        const lineItemNames = order.lineItems.map((item) => getLocalizedStringAsText(item.name));
        const customLineItemNames =
            order.customLineItems?.map((item) => getLocalizedStringAsText(item.name)) || [];
        return Array.from(new Set(lineItemNames.concat(customLineItemNames)));
    }

    private getCategoryNamesFromProduct(categories: CategoryReference[]): string[] {
        const categoryNames = categories.map((category) => {
            const categoryAncestorNames = (category.obj as Category).ancestors.map((ancestor) =>
                getLocalizedStringAsText((ancestor.obj as Category).name),
            );
            const categoryName = getLocalizedStringAsText((category.obj as Category).name);
            const discreteCategoryNames = categoryAncestorNames.concat(categoryName);
            return discreteCategoryNames.map((name, index) => {
                const ancestorNames = [];
                for (let i = index; i >= 0; i--) {
                    ancestorNames.push(discreteCategoryNames[i]);
                }
                return ancestorNames.reverse().join(' > ');
            });
        });

        // Conversion from Array -> Set -> Array removes duplicate entries.
        // Useful for certain cases where you have a common main category such as:
        // Women > Pants > Jeans and Women > Seasonal Fashion > Fall
        return Array.from(new Set(categoryNames.flat()));
    }
}
