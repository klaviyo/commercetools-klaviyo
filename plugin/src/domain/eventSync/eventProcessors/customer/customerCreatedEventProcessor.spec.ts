import { CustomerCreatedEventProcessor } from './customerCreatedEventProcessor';
import { MessageDeliveryPayload } from '@commercetools/platform-sdk/dist/declarations/src/generated/models/subscription';
import { DeepMockProxy, mockDeep } from 'jest-mock-extended';
import { expect as exp } from 'chai';
import { getSampleCustomerCreatedMessage } from '../../../../test/testData/ctCustomerMessages';
import { Context } from '../../../../types/klaviyo-context';
import { DefaultCustomerMapper } from '../../../shared/mappers/DefaultCustomerMapper';
import { getSampleCustomerApiResponse } from '../../../../test/testData/ctCustomerApi';

jest.mock('../../../../infrastructure/driven/klaviyo/KlaviyoService');
const contextMock: DeepMockProxy<Context> = mockDeep<Context>();

contextMock.customerMapper.mapCtCustomerToKlaviyoProfile.mockImplementation(
    (customer, klaviyoProfileId) => (new DefaultCustomerMapper()).mapCtCustomerToKlaviyoProfile(customer, klaviyoProfileId),
);


describe('customerCreatedEvent > isEventValid', () => {
    beforeEach(() => {
        contextMock.klaviyoService.getKlaviyoProfileByExternalId.mockResolvedValue(undefined);
    });

    it('should return valid when the customer created resource has all the required fields', async () => {
        const ctMessageMock: MessageDeliveryPayload = mockDeep<MessageDeliveryPayload>();
        Object.defineProperty(ctMessageMock, 'resource', { value: { typeId: 'customer' } }); //mock readonly property
        Object.defineProperty(ctMessageMock, 'type', { value: 'CustomerCreated' });
        Object.defineProperty(ctMessageMock, 'customer', { value: { email: 'some@email.com' } });

        const event = CustomerCreatedEventProcessor.instance(ctMessageMock, contextMock);

        exp(event.isEventValid()).to.be.true;
    });

    it.each`
        resource      | type                 | customer
        ${'invalid'}  | ${'CustomerCreated'} | ${{ email: 'someemail@klaviyo.com' }}
        ${'customer'} | ${'invalid'}         | ${{ email: 'someemail@klaviyo.com' }}
        ${'customer'} | ${'CustomerCreated'} | ${null}
    `('should return invalid when is not a valid customer created message', ({ resource, type, customer }) => {
        const ctMessageMock: MessageDeliveryPayload = mockDeep<MessageDeliveryPayload>();
        Object.defineProperty(ctMessageMock, 'resource', { value: { typeId: resource } });
        Object.defineProperty(ctMessageMock, 'type', { value: type });
        Object.defineProperty(ctMessageMock, 'customer', { value: customer });

        const event = CustomerCreatedEventProcessor.instance(ctMessageMock, contextMock);

        exp(event.isEventValid()).to.be.false;
    });
});

describe('customerCreatedEvent > generateKlaviyoEvent', () => {
    beforeEach(() => {
        contextMock.klaviyoService.getKlaviyoProfileByExternalId.mockResolvedValue(undefined);
        //todo add tests with profile available
    });
    it('should generate the klaviyo create profile event when the input customer created event is valid', async () => {
        const message = getSampleCustomerCreatedMessage();
        message.customer.addresses.splice(0, message.customer.addresses.length);
        message.customer.addresses.push({
            id: '1235aa3a-5417-4b51-a76c-d6721472531f',
            region: 'aRegion',
            city: 'London',
            country: 'UK',
            phone: '+4407476588266',
            postalCode: 'WE1 2DP',
            streetName: 'High Road',
            streetNumber: '23',
            additionalStreetInfo: 'private access',
            building: 'Tall Tower',
            apartment: 'C',
            additionalAddressInfo: 'additional address info',
            state: 'a state',
        });
        const event = CustomerCreatedEventProcessor.instance(message as unknown as MessageDeliveryPayload, contextMock);

        const klaviyoEvent = await event.generateKlaviyoEvents();

        exp(klaviyoEvent).to.not.be.undefined;
        exp(klaviyoEvent.length).to.be.eq(1);
        expect(contextMock.customerMapper.mapCtCustomerToKlaviyoProfile).toBeCalledTimes(1);
        expect(contextMock.customerMapper.mapCtCustomerToKlaviyoProfile).toBeCalledWith(message.customer);
        expect(klaviyoEvent[0]).toMatchSnapshot();
    });

    it('should generate the klaviyo create profile event getting the customer from CT when is not included in the message', async () => {
        const message: any = {
            ...getSampleCustomerCreatedMessage(),
        };
        delete message.customer;
        contextMock.ctCustomerService.getCustomerProfile.mockResolvedValue(getSampleCustomerApiResponse());
        const event = CustomerCreatedEventProcessor.instance(message as unknown as MessageDeliveryPayload, contextMock);

        const klaviyoEvent = await event.generateKlaviyoEvents();

        exp(klaviyoEvent).to.not.be.undefined;
        exp(klaviyoEvent.length).to.be.eq(1);
        expect(contextMock.customerMapper.mapCtCustomerToKlaviyoProfile).toBeCalledTimes(1);
        expect(contextMock.customerMapper.mapCtCustomerToKlaviyoProfile).toBeCalledWith(getSampleCustomerApiResponse());
        expect(contextMock.ctCustomerService.getCustomerProfile).toBeCalledTimes(1);
        expect(klaviyoEvent[0]).toMatchSnapshot();
    });

    it('should generate the klaviyo create profile event with the default billing address when is available', async () => {
        const message = getSampleCustomerCreatedMessage();
        message.customer.addresses.splice(0, message.customer.addresses.length);
        message.customer.addresses.push({
            id: 'delivery-address-id',
            region: 'aRegion',
            city: 'London',
            country: 'UK',
            phone: '+4407476588266',
            postalCode: 'SE1 5XG',
            streetName: 'some road',
            streetNumber: '23',
            additionalStreetInfo: 'private access',
            building: 'Tall Tower',
            apartment: 'C',
            additionalAddressInfo: 'additional address info',
            state: 'a state',
        });
        message.customer.addresses.push({
            id: 'billing-address-id',
            region: 'aRegion',
            city: 'London',
            country: 'UK',
            phone: '+4407476588266',
            postalCode: 'WS2 5FX',
            streetName: "I'm the default billing address",
            streetNumber: '23',
            additionalStreetInfo: 'private access',
            building: 'Tall Tower',
            apartment: 'C',
            additionalAddressInfo: 'additional address info',
            state: 'a state',
        });
        Object.defineProperty(message, 'customer', {
            value: { ...message.customer, defaultBillingAddressId: 'billing-address-id' },
        }); //mock readonly property
        const event = CustomerCreatedEventProcessor.instance(message as unknown as MessageDeliveryPayload, contextMock);

        const klaviyoEvent = await event.generateKlaviyoEvents();

        exp(message.customer.addresses.length).to.equal(2);
        exp(klaviyoEvent).to.not.be.undefined;
        exp(klaviyoEvent.length).to.be.eq(1);
        expect(contextMock.customerMapper.mapCtCustomerToKlaviyoProfile).toBeCalledTimes(1);
        expect(contextMock.customerMapper.mapCtCustomerToKlaviyoProfile).toBeCalledWith(message.customer);
        expect(klaviyoEvent[0]).toMatchSnapshot();
    });

    it('should generate the klaviyo create profile event with the first billing address available when no default billing address is set', async () => {
        const message = getSampleCustomerCreatedMessage();
        message.customer.addresses.splice(0, message.customer.addresses.length);
        message.customer.addresses.push({
            id: 'delivery-address-id',
            region: 'aRegion',
            city: 'London',
            country: 'UK',
            phone: '+4407476588266',
            postalCode: 'SE1 5XG',
            streetName: 'some road',
            streetNumber: '23',
            additionalStreetInfo: 'private access',
            building: 'Tall Tower',
            apartment: 'C',
            additionalAddressInfo: 'additional address info',
            state: 'a state',
        });
        message.customer.addresses.push({
            id: 'first-billing-address-id',
            region: 'aRegion',
            city: 'London',
            country: 'UK',
            phone: '+4407476588266',
            postalCode: 'WS2 5FX',
            streetName: "I'm a billing address",
            streetNumber: '23',
            additionalStreetInfo: 'private access',
            building: 'Tall Tower',
            apartment: 'C',
            additionalAddressInfo: "I'm a billing address",
            state: 'a state',
        });
        message.customer.addresses.push({
            id: 'second-billing-address-id',
            region: 'aRegion',
            city: 'London',
            country: 'UK',
            phone: '+4407476588266',
            postalCode: 'WS2 5FY',
            streetName: "I'm another billing address",
            streetNumber: '23',
            additionalStreetInfo: 'private access',
            building: 'Tall Tower',
            apartment: 'C',
            additionalAddressInfo: "I'm another billing address",
            state: 'a state',
        });

        Object.defineProperty(message, 'customer', {
            value: {
                ...message.customer,
                billingAddressIds: ['second-billing-address-id', 'first-billing-address-id'],
                defaultBillingAddressId: null,
            },
        }); //mock readonly property

        const event = CustomerCreatedEventProcessor.instance(message as unknown as MessageDeliveryPayload, contextMock);

        const klaviyoEvent = await event.generateKlaviyoEvents();

        exp(message.customer.addresses.length).to.equal(3);
        exp(klaviyoEvent).to.not.be.undefined;
        exp(klaviyoEvent.length).to.be.eq(1);
        expect(contextMock.customerMapper.mapCtCustomerToKlaviyoProfile).toBeCalledTimes(1);
        expect(contextMock.customerMapper.mapCtCustomerToKlaviyoProfile).toBeCalledWith(message.customer);
        expect(klaviyoEvent[0]).toMatchSnapshot();
    });

    it('should generate the klaviyo create profile event with the first address when no billing address ids are set', async () => {
        const message = getSampleCustomerCreatedMessage();
        message.customer.addresses.splice(0, message.customer.addresses.length);
        message.customer.addresses.push({
            id: 'some-address-id',
            region: 'aRegion',
            city: 'London',
            country: 'UK',
            phone: '+4407476588266',
            postalCode: 'WS2 5FX',
            streetName: "I'm a random address",
            streetNumber: '23',
            additionalStreetInfo: 'private access',
            building: 'Tall Tower',
            apartment: 'C',
            additionalAddressInfo: "I'm a random address",
            state: 'a state',
        });
        message.customer.addresses.push({
            id: 'another-address-id',
            region: 'aRegion',
            city: 'London',
            country: 'UK',
            phone: '+4407476588266',
            postalCode: 'WS2 5FY',
            streetName: "I'm another random address",
            streetNumber: '23',
            additionalStreetInfo: 'private access',
            building: 'Tall Tower',
            apartment: 'C',
            additionalAddressInfo: "I'm another random address",
            state: 'a state',
        });

        Object.defineProperty(message, 'customer', {
            value: {
                ...message.customer,
                billingAddressIds: null,
                defaultBillingAddressId: null,
            },
        }); //mock readonly property

        const event = CustomerCreatedEventProcessor.instance(message as unknown as MessageDeliveryPayload, contextMock);

        const klaviyoEvent = await event.generateKlaviyoEvents();

        exp(message.customer.addresses.length).to.equal(2);
        exp(klaviyoEvent).to.not.be.undefined;
        exp(klaviyoEvent.length).to.be.eq(1);
        expect(contextMock.customerMapper.mapCtCustomerToKlaviyoProfile).toBeCalledTimes(1);
        expect(contextMock.customerMapper.mapCtCustomerToKlaviyoProfile).toBeCalledWith(message.customer);
        expect(klaviyoEvent[0]).toMatchSnapshot();
    });

    it('should generate the klaviyo create profile event without customer address inf when no address is available', async () => {
        const message = getSampleCustomerCreatedMessage();
        message.customer.addresses.splice(0, message.customer.addresses.length);
        Object.defineProperty(message, 'customer', {
            value: {
                ...message.customer,
                billingAddressIds: null,
                defaultBillingAddressId: null,
            },
        }); //mock readonly property

        const event = CustomerCreatedEventProcessor.instance(message as unknown as MessageDeliveryPayload, contextMock);

        const klaviyoEvent = await event.generateKlaviyoEvents();

        exp(message.customer.addresses.length).to.equal(0);
        exp(klaviyoEvent).to.not.be.undefined;
        exp(klaviyoEvent.length).to.be.eq(1);
        expect(contextMock.customerMapper.mapCtCustomerToKlaviyoProfile).toBeCalledTimes(1);
        expect(contextMock.customerMapper.mapCtCustomerToKlaviyoProfile).toBeCalledWith(message.customer);
        expect(klaviyoEvent[0]).toMatchSnapshot();
    });
});
