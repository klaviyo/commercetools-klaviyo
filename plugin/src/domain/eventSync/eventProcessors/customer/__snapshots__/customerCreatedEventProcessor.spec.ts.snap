// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`customerCreatedEvent > generateKlaviyoEvent should generate the klaviyo create profile event getting the customer from CT when is not included in the message 1`] = `
{
  "body": {
    "data": {
      "attributes": {
        "email": "Alfred_Lebsack@hotmail.com",
        "external_id": "77d8e6f4-2dee-4ac9-bd21-4035b6ba76d8",
        "first_name": "Jailyn",
        "last_name": "Hilll",
        "location": {
          "address1": "12, First Street",
          "address2": "",
          "city": "Brookefort",
          "country": "NL",
          "region": undefined,
          "zip": "12345",
        },
        "organization": undefined,
        "phone_number": "+39 3234567892",
        "properties": undefined,
        "title": "Miss",
      },
      "id": undefined,
      "type": "profile",
    },
  },
  "type": "profileCreated",
}
`;

exports[`customerCreatedEvent > generateKlaviyoEvent should generate the klaviyo create profile event when the input customer created event is valid 1`] = `
{
  "body": {
    "data": {
      "attributes": {
        "email": "roberto.smith@klaviyo.com",
        "external_id": "e54d8233-be41-4ce0-ae68-5d0674dd8517",
        "first_name": "Roberto",
        "last_name": "Smith",
        "location": {
          "address1": "C, Tall Tower, 23, High Road",
          "address2": "private access, additional address info",
          "city": "London",
          "country": "UK",
          "region": "aRegion",
          "zip": "WE1 2DP",
        },
        "organization": "Klaviyo",
        "phone_number": "+4407476588266",
        "properties": undefined,
        "title": "Mr",
      },
      "id": undefined,
      "type": "profile",
    },
  },
  "type": "profileCreated",
}
`;

exports[`customerCreatedEvent > generateKlaviyoEvent should generate the klaviyo create profile event with the default billing address when is available 1`] = `
{
  "body": {
    "data": {
      "attributes": {
        "email": "roberto.smith@klaviyo.com",
        "external_id": "e54d8233-be41-4ce0-ae68-5d0674dd8517",
        "first_name": "Roberto",
        "last_name": "Smith",
        "location": {
          "address1": "C, Tall Tower, 23, I'm the default billing address",
          "address2": "private access, additional address info",
          "city": "London",
          "country": "UK",
          "region": "aRegion",
          "zip": "WS2 5FX",
        },
        "organization": "Klaviyo",
        "phone_number": "+4407476588266",
        "properties": undefined,
        "title": "Mr",
      },
      "id": undefined,
      "type": "profile",
    },
  },
  "type": "profileCreated",
}
`;

exports[`customerCreatedEvent > generateKlaviyoEvent should generate the klaviyo create profile event with the first address when no billing address ids are set 1`] = `
{
  "body": {
    "data": {
      "attributes": {
        "email": "roberto.smith@klaviyo.com",
        "external_id": "e54d8233-be41-4ce0-ae68-5d0674dd8517",
        "first_name": "Roberto",
        "last_name": "Smith",
        "location": {
          "address1": "C, Tall Tower, 23, I'm a random address",
          "address2": "private access, I'm a random address",
          "city": "London",
          "country": "UK",
          "region": "aRegion",
          "zip": "WS2 5FX",
        },
        "organization": "Klaviyo",
        "phone_number": "+4407476588266",
        "properties": undefined,
        "title": "Mr",
      },
      "id": undefined,
      "type": "profile",
    },
  },
  "type": "profileCreated",
}
`;

exports[`customerCreatedEvent > generateKlaviyoEvent should generate the klaviyo create profile event with the first billing address available when no default billing address is set 1`] = `
{
  "body": {
    "data": {
      "attributes": {
        "email": "roberto.smith@klaviyo.com",
        "external_id": "e54d8233-be41-4ce0-ae68-5d0674dd8517",
        "first_name": "Roberto",
        "last_name": "Smith",
        "location": {
          "address1": "C, Tall Tower, 23, I'm another billing address",
          "address2": "private access, I'm another billing address",
          "city": "London",
          "country": "UK",
          "region": "aRegion",
          "zip": "WS2 5FY",
        },
        "organization": "Klaviyo",
        "phone_number": "+4407476588266",
        "properties": undefined,
        "title": "Mr",
      },
      "id": undefined,
      "type": "profile",
    },
  },
  "type": "profileCreated",
}
`;

exports[`customerCreatedEvent > generateKlaviyoEvent should generate the klaviyo create profile event without customer address inf when no address is available 1`] = `
{
  "body": {
    "data": {
      "attributes": {
        "email": "roberto.smith@klaviyo.com",
        "external_id": "e54d8233-be41-4ce0-ae68-5d0674dd8517",
        "first_name": "Roberto",
        "last_name": "Smith",
        "location": null,
        "organization": "Klaviyo",
        "phone_number": undefined,
        "properties": undefined,
        "title": "Mr",
      },
      "id": undefined,
      "type": "profile",
    },
  },
  "type": "profileCreated",
}
`;
